File: scripts/crosslink-artists.ts
#!/usr/bin/env node

/**
 * Crosslinking Script: Links Artists and Knowledge pages bidirectionally
 *
 * Features:
 * - Knowledge → Artists: Adds markdown links in content when artist name is found
 * - Artist → Knowledge: Adds knowledge articles to relatedArticles array
 * - Dry-run mode: Preview changes without modifying files
 * - Backup: Creates .backup files before modifications
 * - Smart matching: Whole-word matching with word boundaries
 */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

// Types
interface Artist {
  id: string;
  slug: string;
  name: string;
  filePath: string;
  frontmatter: Record<string, unknown>;
}

interface Knowledge {
  id: string;
  slug: string;
  title: string;
  filePath: string;
  content: string;
  frontmatter: Record<string, unknown>;
}

interface CrosslinkResult {
  artistPath: string;
  knowledgePath: string;
  matchCount: number;
  linksAdded: string[];
}

interface UpdateResult {
  filePath: string;
  changes: string[];
  originalContent: string;
  newContent: string;
}

// Configuration
const CONFIG = {
  ARTISTS_DIR: path.join(process.cwd(), "src/content/artists"),
  KNOWLEDGE_DIR: path.join(process.cwd(), "src/content/knowledge-en"),
  DRY_RUN: process.argv.includes("--dry-run"),
  BACKUP: !process.argv.includes("--no-backup"),
  FORCE: process.argv.includes("--force"),
};

// Utility: Pure function to read file safely
const readFile = (
  filePath: string
): { success: true; content: string } | { success: false; error: string } => {
  try {
    return { success: true, content: fs.readFileSync(filePath, "utf-8") };
  } catch (error) {
    return {
      success: false,
      error: `Failed to read ${filePath}: ${(error as Error).message}`,
    };
  }
};

// Utility: Pure function to write file safely
const writeFile = (
  filePath: string,
  content: string
): { success: true } | { success: false; error: string } => {
  try {
    fs.writeFileSync(filePath, content, "utf-8");
    return { success: true };
  } catch (error) {
    return {
      success: false,
      error: `Failed to write ${filePath}: ${(error as Error).message}`,
    };
  }
};

// Utility: Atomic file write (write to temp, then rename)
const writeFileAtomic = (
  filePath: string,
  content: string
): { success: true } | { success: false; error: string } => {
  try {
    const tempPath = `${filePath}.tmp`;
    const writeResult = writeFile(tempPath, content);
    if (!writeResult.success) {
      return writeResult;
    }

    fs.renameSync(tempPath, filePath);
    return { success: true };
  } catch (error) {
    return {
      success: false,
      error: `Failed to write ${filePath}: ${(error as Error).message}`,
    };
  }
};

// Utility: Create backup
const createBackup = (
  filePath: string,
  force = false
): { success: true; backupPath: string } | { success: false; error: string } => {
  try {
    const backupPath = `${filePath}.backup`;

    // Check if backup already exists
    if (!force && fs.existsSync(backupPath)) {
      return {
        success: false,
        error: `Backup file already exists at ${backupPath} (use --force to overwrite)`,
      };
    }

    const result = readFile(filePath);
    if (!result.success) {
      return { success: false, error: result.error };
    }
    const writeResult = writeFile(backupPath, result.content);
    if (!writeResult.success) {
      return { success: false, error: writeResult.error };
    }
    return { success: true, backupPath };
  } catch (error) {
    return {
      success: false,
      error: `Failed to create backup: ${(error as Error).message}`,
    };
  }
};

// Pure function: Extract YAML frontmatter from markdown
const parseFrontmatter = (
  markdown: string
): { frontmatter: Record<string, unknown>; content: string } => {
  const frontmatterRegex = /^---\n([\s\S]*?)\n---\n([\s\S]*)$/;
  const match = markdown.match(frontmatterRegex);

  if (!match) {
    return { frontmatter: {}, content: markdown };
  }

  // Simple YAML parsing (for basic key-value pairs and arrays)
  const yamlLines = match[1].split("\n");
  const frontmatter: Record<string, unknown> = {};
  let inArray = false;
  let currentArray: string[] = [];
  let currentKey = "";
  let expectingValueOnNextLine = false;

  yamlLines.forEach((line, index) => {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) {
      // Skip empty lines and comments
      return;
    }

    // Handle array items
    if (trimmed.startsWith("- ")) {
      inArray = true;
      currentArray.push(trimmed.slice(2).replace(/^["']|["']$/g, ""));
      return;
    }

    // Handle key-value pairs
    if (trimmed.includes(":")) {
      // If we were expecting a value on next line, save it first
      if (expectingValueOnNextLine && currentKey) {
        // Value was on previous lines (array)
        frontmatter[currentKey] = currentArray;
        inArray = false;
        currentArray = [];
      }

      const [key, ...valueParts] = trimmed.split(":");
      if (key) {
        currentKey = key.trim();

        // Check if value is on the same line
        if (valueParts.length > 0) {
          const value = valueParts
            .join(":")
            .trim()
            .replace(/^["']|["']$/g, "");
          frontmatter[currentKey] = value;
          expectingValueOnNextLine = false;
        } else {
          // Value is on next line(s)
          expectingValueOnNextLine = true;
          inArray = false;
          currentArray = [];
        }
      }
    } else if (expectingValueOnNextLine && currentKey) {
      // Value line for block scalar
      const value = trimmed.replace(/^["']|["']$/g, "");
      frontmatter[currentKey] = value;
      expectingValueOnNextLine = false;
    }
  });

  // Don't forget the last array
  if (inArray && currentKey) {
    frontmatter[currentKey] = currentArray;
  }

  return { frontmatter, content: match[2] || "" };
};

// Pure function: Serialize object to YAML
const toYaml = (obj: Record<string, unknown>, indent = 0): string => {
  const indentStr = "  ".repeat(indent);
  let yaml = "";

  Object.entries(obj).forEach(([key, value]) => {
    if (Array.isArray(value)) {
      yaml += `${indentStr}${key}:\n`;
      value.forEach((item) => {
        yaml += `${indentStr}  - "${item}"\n`;
      });
    } else if (value === null || value === undefined) {
      yaml += `${indentStr}${key}: null\n`;
    } else if (typeof value === "string") {
      yaml += `${indentStr}${key}: "${value}"\n`;
    } else {
      yaml += `${indentStr}${key}: ${String(value)}\n`;
    }
  });

  return yaml;
};

// Pure function: Rebuild markdown with updated frontmatter
const rebuildMarkdown = (
  frontmatter: Record<string, unknown>,
  content: string
): string => {
  return `---\n${toYaml(frontmatter)}---\n${content}`;
};

// Pure function: Scan artists directory
const scanArtists = (): Artist[] => {
  const files = fs.readdirSync(CONFIG.ARTISTS_DIR).filter((f) => f.endsWith(".md"));

  return files
    .map((file) => {
      const filePath = path.join(CONFIG.ARTISTS_DIR, file);
      const result = readFile(filePath);
      const { frontmatter } = result.success
        ? parseFrontmatter(result.content)
        : { frontmatter: {} };

      const name = (frontmatter as Record<string, unknown>).name as string;

      // Validate required fields
      if (!name || typeof name !== "string") {
        console.warn(`⚠️  Missing or invalid 'name' field in ${file}`);
      }

      return {
        id: file.replace(".md", ""),
        slug: file.replace(".md", ""),
        name: name || "",
        filePath,
        frontmatter,
      };
    })
    .filter((artist) => artist.name.length > 0); // Filter out invalid artists
};

// Pure function: Scan knowledge directory
const scanKnowledge = (): Knowledge[] => {
  const files = fs
    .readdirSync(CONFIG.KNOWLEDGE_DIR)
    .filter((f) => f.endsWith(".md") || f.endsWith(".mdx"));

  return files
    .map((file) => {
      const filePath = path.join(CONFIG.KNOWLEDGE_DIR, file);
      const result = readFile(filePath);
      const { frontmatter, content } = result.success
        ? parseFrontmatter(result.content)
        : { frontmatter: {}, content: "" };

      const title = (frontmatter as Record<string, unknown>).title as string;

      // Validate required fields
      if (!title || typeof title !== "string") {
        console.warn(`⚠️  Missing or invalid 'title' field in ${file}`);
      }

      return {
        id: file,
        slug: file.replace(/\.(md|mdx)$/, ""),
        title: title || "",
        filePath,
        content,
        frontmatter,
      };
    })
    .filter((knowledge) => knowledge.title.length > 0); // Filter out invalid knowledge pages
};

// Pure function: Check if artist name exists in knowledge content
const findArtistMatches = (artist: Artist, knowledge: Knowledge): string[] => {
  const matches: string[] = [];
  const name = artist.name;
  const content = knowledge.content;

  // Create regex pattern with word boundaries
  const pattern = new RegExp(`\\b${escapeRegex(name)}\\b`, "g");

  let match;
  while ((match = pattern.exec(content)) !== null) {
    // Check if this is already a markdown link
    const before = content.substring(Math.max(0, match.index - 1), match.index);
    const after = content.substring(
      match.index + match[0].length,
      match.index + match[0].length + 1
    );

    if (before !== "[" && after !== ")") {
      matches.push(match[0]);
    }
  }

  return matches;
};

// Helper: Escape regex special characters
const escapeRegex = (str: string): string => str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

// Pure function: Add markdown links for artist mentions
const addArtistLinksToKnowledge = (
  artist: Artist,
  knowledge: Knowledge
): { content: string; linksAdded: number } => {
  let content = knowledge.content;
  let linksAdded = 0;
  const name = artist.name;
  const url = `/artists/${artist.slug}`;

  // Replace whole-word matches that are not already links
  // Note: We use a simpler pattern without lookbehind for better compatibility
  const pattern = new RegExp(`\\b${escapeRegex(name)}\\b`, "g");

  content = content.replace(pattern, (match, offset) => {
    // Check if this is already part of a markdown link
    const before = content.substring(Math.max(0, offset - 1), offset);
    const after = content.substring(offset + match.length, offset + match.length + 1);

    if (before !== "[" && after !== ")") {
      linksAdded++;
      return `[${match}](${url})`;
    }
    return match;
  });

  return { content, linksAdded };
};

// Pure function: Check if knowledge already exists in artist's relatedArticles
const hasRelatedArticle = (artist: Artist, knowledgeSlug: string): boolean => {
  const relatedArticles = (artist.frontmatter.relatedArticles as string[]) || [];
  return relatedArticles.includes(knowledgeSlug);
};

// Pure function: Add knowledge to artist's relatedArticles
const addKnowledgeToArtist = (artist: Artist, knowledgeSlug: string): Artist => {
  const relatedArticles = (artist.frontmatter.relatedArticles as string[]) || [];

  return {
    ...artist,
    frontmatter: {
      ...artist.frontmatter,
      relatedArticles: [...relatedArticles, knowledgeSlug],
    },
  };
};

// Main crosslinking logic
const runCrosslinking = (): {
  knowledgeUpdates: UpdateResult[];
  artistUpdates: UpdateResult[];
  summary: { knowledgeLinksAdded: number; artistLinksAdded: number };
} => {
  const artists = scanArtists();
  const knowledgePages = scanKnowledge();
  const knowledgeUpdates: UpdateResult[] = [];
  const artistUpdates: UpdateResult[] = [];

  console.log(
    `\n🔍 Scanning ${artists.length} artists and ${knowledgePages.length} knowledge pages...\n`
  );

  // Knowledge → Artists: Add markdown links
  knowledgePages.forEach((knowledge) => {
    let updatedContent = knowledge.content;
    const changes: string[] = [];

    artists.forEach((artist) => {
      const result = addArtistLinksToKnowledge(artist, {
        ...knowledge,
        content: updatedContent,
      });
      if (result.linksAdded > 0) {
        updatedContent = result.content;
        changes.push(`Added ${result.linksAdded} link(s) to ${artist.name}`);
      }
    });

    if (changes.length > 0) {
      const originalReadResult = readFile(knowledge.filePath);
      knowledgeUpdates.push({
        filePath: knowledge.filePath,
        changes,
        originalContent: originalReadResult.success ? originalReadResult.content : "",
        newContent: rebuildMarkdown(knowledge.frontmatter, updatedContent),
      });
    }
  });

  // Artist → Knowledge: Add to relatedArticles
  artists.forEach((artist) => {
    const changes: string[] = [];
    const content = readFile(artist.filePath);

    if (!content.success) {
      console.warn(`⚠️  Skipping ${artist.filePath}: ${content.error}`);
      return;
    }

    const { frontmatter, content: markdownContent } = parseFrontmatter(content.content);
    let relatedArticles = (frontmatter.relatedArticles as string[]) || [];

    knowledgePages.forEach((knowledge) => {
      if (!hasRelatedArticle(artist, knowledge.slug)) {
        // Check if artist is mentioned in knowledge
        const matches = findArtistMatches(artist, knowledge);
        if (matches.length > 0) {
          relatedArticles.push(knowledge.slug);
          changes.push(`Added related article: ${knowledge.title}`);
        }
      }
    });

    if (changes.length > 0) {
      artistUpdates.push({
        filePath: artist.filePath,
        changes,
        originalContent: content.content,
        newContent: rebuildMarkdown({ ...frontmatter, relatedArticles }, markdownContent),
      });
    }
  });

  const summary = {
    knowledgeLinksAdded: knowledgeUpdates.reduce((sum, u) => sum + u.changes.length, 0),
    artistLinksAdded: artistUpdates.reduce((sum, u) => sum + u.changes.length, 0),
  };

  return { knowledgeUpdates, artistUpdates, summary };
};

// Apply updates to files
const applyUpdates = (updates: UpdateResult[], type: string): void => {
  updates.forEach((update) => {
    console.log(`\n📄 ${type}: ${path.basename(update.filePath)}`);
    update.changes.forEach((change) => {
      console.log(`   ✅ ${change}`);
    });

    if (CONFIG.DRY_RUN) {
      console.log(`   ℹ️  (Dry-run: Not modifying file)`);
      return;
    }

    if (CONFIG.BACKUP) {
      const backupResult = createBackup(update.filePath, CONFIG.FORCE);
      if (!backupResult.success) {
        console.log(`   ❌ Backup failed: ${backupResult.error}`);
        return;
      }
      console.log(`   💾 Backup created: ${path.basename(backupResult.backupPath)}`);
    }

    // Use atomic write for better safety
    const writeResult = writeFileAtomic(update.filePath, update.newContent);
    if (writeResult.success) {
      console.log(`   ✍️  File updated`);
    } else {
      console.log(`   ❌ Failed: ${writeResult.error}`);
    }
  });
};

// Main execution
const main = (): void => {
  console.log("\n🔗 Crosslinking Script: Artists ↔ Knowledge");
  console.log("=".repeat(50));
  console.log(`Mode: ${CONFIG.DRY_RUN ? "DRY-RUN" : "LIVE"}`);
  console.log(`Backup: ${CONFIG.BACKUP ? "enabled" : "disabled"}`);
  console.log(`Force backup overwrite: ${CONFIG.FORCE ? "enabled" : "disabled"}\n`);

  const { knowledgeUpdates, artistUpdates, summary } = runCrosslinking();

  console.log("\n" + "=".repeat(50));
  console.log("📊 SUMMARY");
  console.log("=".repeat(50));
  console.log(`Knowledge pages to update: ${knowledgeUpdates.length}`);
  console.log(`Artist pages to update: ${artistUpdates.length}`);
  console.log(`Total artist links to add: ${summary.knowledgeLinksAdded}`);
  console.log(`Total related articles to add: ${summary.artistLinksAdded}`);

  if (knowledgeUpdates.length > 0 || artistUpdates.length > 0) {
    console.log("\n" + "=".repeat(50));
    console.log("📝 CHANGES");
    console.log("=".repeat(50));
    applyUpdates(knowledgeUpdates, "Knowledge");
    applyUpdates(artistUpdates, "Artist");
  } else {
    console.log("\n✨ No changes needed - everything is already crosslinked!");
  }

  console.log("\n" + "=".repeat(50));
  console.log("✅ Done!\n");
};

// Run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

export { main as crosslink };

File: scripts/crosslink-artists.py
#!/usr/bin/env python3

"""
Crosslink artists and knowledge pages.

Behavior:
- Knowledge -> Artists: add markdown links for artist mentions in body content.
- Artists -> Knowledge: strict-sync relatedArticles to actual mentions only.

Safety:
- --dry-run to preview without writing.
- backup files (.backup) before each write, unless --no-backup.
- atomic writes via temporary file + replace.
"""

from __future__ import annotations

import argparse
import re
import shutil
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

try:
    import yaml
except ImportError:
    print("Error: missing dependency pyyaml. Install with: pip install pyyaml")
    sys.exit(1)


ROOT = Path.cwd()
ARTISTS_DIR = ROOT / "src/content/artists"
KNOWLEDGE_DIR = ROOT / "src/content/knowledge-en"

FRONTMATTER_PATTERN = re.compile(
    r"^---\r?\n([\s\S]*?)\r?\n---\r?\n([\s\S]*)$",
    re.DOTALL,
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Crosslink artists and knowledge content")
    parser.add_argument("--dry-run", action="store_true", help="Preview changes only")
    parser.add_argument("--no-backup", action="store_true", help="Skip creating backups")
    parser.add_argument("--force", action="store_true", help="Overwrite existing backups")
    return parser.parse_args()


def read_text(path: Path) -> Optional[str]:
    try:
        return path.read_text(encoding="utf-8")
    except Exception as exc:
        print(f"WARN: failed to read {path}: {exc}")
        return None


def parse_frontmatter(content: str) -> Tuple[Optional[Dict[str, Any]], str]:
    match = FRONTMATTER_PATTERN.match(content)
    if not match:
        return None, content

    frontmatter_raw, body = match.group(1), match.group(2)
    try:
        data = yaml.safe_load(frontmatter_raw)
    except yaml.YAMLError as exc:
        print(f"WARN: yaml parse failed: {exc}")
        return None, body

    if not isinstance(data, dict):
        return None, body
    return data, body


def split_frontmatter_raw(content: str) -> Tuple[Optional[str], str]:
    match = FRONTMATTER_PATTERN.match(content)
    if not match:
        return None, content
    return match.group(1), match.group(2)


def replace_body_preserve_frontmatter(content: str, new_body: str) -> str:
    raw_frontmatter, _ = split_frontmatter_raw(content)
    if raw_frontmatter is None:
        return content
    return f"---\n{raw_frontmatter}\n---\n{new_body}"


def upsert_related_articles_preserve_frontmatter(content: str, slugs: List[str]) -> str:
    raw_frontmatter, body = split_frontmatter_raw(content)
    if raw_frontmatter is None:
        return content

    pattern = re.compile(r"(?m)^relatedArticles:\r?\n(?:[ \t]*-[^\n]*\r?\n?)*")

    if slugs:
        block = "relatedArticles:\n" + "\n".join(f'  - "{slug}"' for slug in slugs) + "\n"
    else:
        block = ""

    if pattern.search(raw_frontmatter):
        updated_frontmatter = pattern.sub(block, raw_frontmatter)
    else:
        if not block:
            updated_frontmatter = raw_frontmatter
        else:
            sep = "" if raw_frontmatter.endswith("\n") else "\n"
            updated_frontmatter = f"{raw_frontmatter}{sep}{block}"

    return f"---\n{updated_frontmatter}\n---\n{body}"


def build_protected_ranges(text: str) -> List[Tuple[int, int]]:
    ranges: List[Tuple[int, int]] = []
    for m in re.finditer(r"```[\s\S]*?```", text):
        ranges.append((m.start(), m.end()))
    for m in re.finditer(r"`[^`\n]+`", text):
        ranges.append((m.start(), m.end()))
    return ranges


def in_ranges(pos: int, ranges: List[Tuple[int, int]]) -> bool:
    for start, end in ranges:
        if start <= pos < end:
            return True
    return False


def create_backup(path: Path, force: bool) -> Tuple[bool, str]:
    backup = Path(f"{path}.backup")
    if backup.exists() and not force:
        return False, f"backup exists: {backup.name} (use --force)"
    try:
        shutil.copy2(path, backup)
        return True, f"backup created: {backup.name}"
    except Exception as exc:
        return False, f"backup failed: {exc}"


def write_atomic(path: Path, content: str) -> Tuple[bool, str]:
    tmp = path.with_suffix(f"{path.suffix}.tmp")
    try:
        tmp.write_text(content, encoding="utf-8")
        tmp.replace(path)
        return True, "ok"
    except Exception as exc:
        return False, f"write failed: {exc}"


def scan_artists() -> List[Dict[str, Any]]:
    artists: List[Dict[str, Any]] = []
    for path in sorted(ARTISTS_DIR.glob("*.md")):
        content = read_text(path)
        if content is None:
            continue
        fm, _body = parse_frontmatter(content)
        if not fm:
            print(f"WARN: skipping {path.name}: no valid frontmatter")
            continue
        name = fm.get("name")
        if not isinstance(name, str) or not name.strip():
            print(f"WARN: skipping {path.name}: missing name")
            continue
        artists.append(
            {
                "slug": path.stem,
                "name": name,
                "path": path,
                "frontmatter": fm,
                "content": content,
            }
        )
    return artists


def parse_knowledge_entry(path: Path) -> Optional[Dict[str, Any]]:
    if path.name.endswith(".backup") or path.name.endswith(".tmp"):
        return None
    content = read_text(path)
    if content is None:
        return None
    fm, body = parse_frontmatter(content)
    if not fm:
        print(f"WARN: skipping {path.name}: no valid frontmatter")
        return None
    title = fm.get("title")
    if not isinstance(title, str) or not title.strip():
        print(f"WARN: skipping {path.name}: missing title")
        return None
    return {
        "slug": path.stem,
        "title": title,
        "path": path,
        "frontmatter": fm,
        "content": content,
        "body": body,
    }


def scan_knowledge() -> List[Dict[str, Any]]:
    entries: List[Dict[str, Any]] = []
    paths = sorted(KNOWLEDGE_DIR.glob("*.mdx")) + sorted(KNOWLEDGE_DIR.glob("*.md"))
    for path in paths:
        entry = parse_knowledge_entry(path)
        if entry is not None:
            entries.append(entry)
    return entries


def add_links_for_artist(name: str, slug: str, body: str) -> Tuple[str, int]:
    protected = build_protected_ranges(body)
    pattern = re.compile(rf"\b{re.escape(name)}\b", re.IGNORECASE)
    url = f"/artists/{slug}"
    added = 0

    def repl(match: re.Match[str]) -> str:
        nonlocal added
        start, end = match.start(), match.end()
        if in_ranges(start, protected):
            return match.group(0)
        before = body[start - 1] if start > 0 else ""
        after = body[end : end + 2]
        if before == "[" and after == "](":
            return match.group(0)
        added += 1
        return f"[{match.group(0)}]({url})"

    return pattern.sub(repl, body), added


def has_mention(name: str, body: str) -> bool:
    protected = build_protected_ranges(body)
    pattern = re.compile(rf"\b{re.escape(name)}\b", re.IGNORECASE)
    for match in pattern.finditer(body):
        if not in_ranges(match.start(), protected):
            return True
    return False


def run_crosslink() -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]], Dict[str, int]]:
    artists = scan_artists()
    knowledge_pages = scan_knowledge()

    knowledge_updates: List[Dict[str, Any]] = []
    artist_updates: List[Dict[str, Any]] = []
    summary = {"knowledgeLinksAdded": 0, "artistLinksAdded": 0}

    print()
    print(f"Scanning {len(artists)} artists and {len(knowledge_pages)} knowledge pages")

    for knowledge in knowledge_pages:
        body = knowledge["body"]
        changes: List[str] = []
        for artist in artists:
            body, added = add_links_for_artist(artist["name"], artist["slug"], body)
            if added > 0:
                changes.append(f"added {added} link(s) to {artist['name']}")
                summary["knowledgeLinksAdded"] += added

        if changes:
            knowledge_updates.append(
                {
                    "path": knowledge["path"],
                    "new_content": replace_body_preserve_frontmatter(knowledge["content"], body),
                    "changes": changes,
                }
            )

    for artist in artists:
        existing = artist["frontmatter"].get("relatedArticles", [])
        if not isinstance(existing, list):
            existing = []

        desired: List[str] = []
        for knowledge in knowledge_pages:
            if has_mention(artist["name"], knowledge["body"]):
                desired.append(knowledge["slug"])

        added = [slug for slug in desired if slug not in existing]
        removed = [slug for slug in existing if slug not in desired]
        if not added and not removed:
            continue

        changes = [f"added related article: {slug}" for slug in added]
        changes.extend(f"removed unrelated article: {slug}" for slug in removed)
        summary["artistLinksAdded"] += len(added)

        artist_updates.append(
            {
                "path": artist["path"],
                "new_content": upsert_related_articles_preserve_frontmatter(
                    artist["content"], desired
                ),
                "changes": changes,
            }
        )

    return knowledge_updates, artist_updates, summary


def apply_updates(
    updates: List[Dict[str, Any]],
    label: str,
    dry_run: bool,
    no_backup: bool,
    force: bool,
) -> None:
    if not updates:
        print(f"No {label.lower()} updates")
        return

    for update in updates:
        path: Path = update["path"]
        print()
        print(f"{label}: {path.name}")
        for change in update["changes"]:
            print(f"  - {change}")

        if dry_run:
            print("  - dry-run: no file changes")
            continue

        if not no_backup:
            ok, msg = create_backup(path, force)
            if not ok:
                print(f"  - backup error: {msg}")
                continue
            print(f"  - {msg}")

        ok, msg = write_atomic(path, update["new_content"])
        if not ok:
            print(f"  - {msg}")
            continue
        print("  - file updated")


def main() -> None:
    args = parse_args()

    print("Crosslink artists <-> knowledge")
    print(f"Mode: {'dry-run' if args.dry_run else 'live'}")
    print(f"Backup: {'disabled' if args.no_backup else 'enabled'}")
    print(f"Force backup overwrite: {'enabled' if args.force else 'disabled'}")

    knowledge_updates, artist_updates, summary = run_crosslink()

    print()
    print("Summary")
    print(f"- Knowledge pages to update: {len(knowledge_updates)}")
    print(f"- Artist pages to update: {len(artist_updates)}")
    print(f"- Total artist links to add: {summary['knowledgeLinksAdded']}")
    print(f"- Total related articles added: {summary['artistLinksAdded']}")

    if not knowledge_updates and not artist_updates:
        print("No changes needed")
        return

    print()
    print("Planned changes")
    apply_updates(knowledge_updates, "Knowledge", args.dry_run, args.no_backup, args.force)
    apply_updates(artist_updates, "Artist", args.dry_run, args.no_backup, args.force)


if __name__ == "__main__":
    main()
